.github/workflows/docker.yml

```diff
@@ -0,0 +1,78 @@
+        run: |
+          docker buildx build . --builder=container --cache-from=type=registry,ref=ghcr.io/datadog/catadog --load --platform linux/x86_64 -f ./Dockerfile --tag ghcr.io/datadog/catadog
+      - name: Test single-arch image (x86-64)
+        run: |
+          docker run --platform linux/x86_64 --rm ghcr.io/datadog/catadog /bin/sh -c 'true'
+          docker run --platform linux/x86_64 --rm ghcr.io/datadog/catadog ruby -e 'puts RUBY_DESCRIPTION'
+          docker run --platform linux/x86_64 --rm ghcr.io/datadog/catadog gem --version
+          docker run --platform linux/x86_64 --rm ghcr.io/datadog/catadog bundle --version
+          docker run --platform linux/x86_64 --rm -v "${PWD}":"${PWD}" -w "${PWD}" ghcr.io/datadog/catadog /bin/sh -c 'bundle install && bundle exec rake test'
+
+      # Build image for aarch64-linux, emulated under qemu
+      # 
+      # Tag image separately to avoid interference with caching and so that testing step can reference the image
+      - name: Enable aarch64 emulation (x86-64)
+        run: |
+          docker run --privileged --rm tonistiigi/binfmt --install arm64
+      - name: Build single-arch image (aarch64-linux)
+        run: |
+          docker buildx build . --builder=container --cache-from=type=registry,ref=ghcr.io/datadog/catadog --output=type=image,push=false --platform linux/aarch64 -f ./Dockerfile
+      - name: Tag single-arch image (aarch64-linux)
+        run: |
+          docker buildx build . --builder=container --cache-from=type=registry,ref=ghcr.io/datadog/catadog --load --platform linux/aarch64 -f ./Dockerfile --tag ghcr.io/datadog/catadog
+      - name: Test single-arch image (aarch64-linux)
+        run: |
+          docker run --platform linux/aarch64 --rm ghcr.io/datadog/catadog /bin/sh -c 'true'
+          docker run --platform linux/aarch64 --rm ghcr.io/datadog/catadog ruby -e 'puts RUBY_DESCRIPTION'
+          docker run --platform linux/aarch64 --rm ghcr.io/datadog/catadog gem --version
+          docker run --platform linux/aarch64 --rm ghcr.io/datadog/catadog bundle --version
+          docker run --platform linux/aarch64 --rm -v "${PWD}":"${PWD}" -w "${PWD}" ghcr.io/datadog/catadog /bin/sh -c 'bundle install && bundle exec rake test'
+      
+      # Assemble multi-arch image for a combined push to the registry
+      # 
+      # Docker build is rerun, but build is fast because the layers are already cached
+      - name: Log in to the container registry
+        if: ${{ inputs.push }}
+        run: |
+          echo ${{ secrets.GITHUB_TOKEN }} | docker login ghcr.io -u ${{ github.actor }} --password-stdin
+      - name: Build multi-arch image (x86-64, aarch64)
+        if: ${{ inputs.push }}
+        run: |
+          docker buildx build . --builder=container --cache-from=type=registry,ref=ghcr.io/datadog/catadog --output=type=image,push=true --build-arg BUILDKIT_INLINE_CACHE=1 --platform linux/x86_64,linux/aarch64 -f ./Dockerfile --tag ghcr.io/datadog/catadog
```

Comment:
The image is currently being pushed as `ghcr.io/datadog/catadog` with no tag. One suggestion was to add the tag based on the `catadog` version, but I'm not sure how to get the version number if there are no official releases. Another suggestion was to pair the image with the `github.run_id`. However, I thought that may be too unique of an identifier, as each workflow run would have a different run_id. The [workflow that this is based on](https://github.com/DataDog/images-rb/blob/main/.github/workflows/build-ruby.yml#L113-L123) tags the image based on the ruby / jruby version. Since this workflow only tests one ruby version (3.4), I did not see the value of adding the 3.4 tag. Should there still be a tag though? and if so, what could that tag be?

Reply:
> tag based on the `catadog` version

Yeah this was my suggestion. I think it would mean storing the version in a plain-text file and sourcing the file to set `catadog`'s version and the image's version. Alternatively, if there was some kind of `gem` or `bundle` command that could parse out the version from the `.gemspec`, that could work too. I'm afraid I don't know what the latter would look off the top of my head.

I don't recommend run ID, mostly because its not meaningfully linked to a particular state of the Ruby gem. At least, it would require some digging through build logs to figure out what version was being built at the time... not a fun experience.

Reply:
> I don't recommend run ID, mostly because its not meaningfully linked to a particular state of the Ruby gem

`run_id` is essential for auditability and reproducibility: two runs of the same commit may produce different results because of environment changes.

Note that `run_id` is used only to use the proper image across, say, a workflow's jobs. Without `run_id` one runs into concurrency issues: imagine two runs are triggered for the same commit and the only identifier is the commid id; since we don't control the scheduling of jobs, if the sole identifier is the commit id there is an opportunity for a run to end up picking the _other_ run's output.

This process is entirely different and decoupled from a _release_ process, where there would be a version, a tag, and a single set of artifacts, which, once published, should be immutable. Nonetheless, the release process can reuse the run artifacts and promote them to be tagged with a simple version instead.